src/sentry/tagstore/snuba/backend.py

```diff
@@ -172,18 +176,34 @@ def __get_tag_keys_for_projects(
             aggregations.append(["uniq", "tags_value", "values_seen"])
         conditions = []
 
-        result = snuba.query(
-            start=start,
-            end=end,
-            groupby=["tags_key"],
-            conditions=conditions,
-            filter_keys=filters,
-            aggregations=aggregations,
-            limit=limit,
-            orderby="-count",
-            referrer="tagstore.__get_tag_keys",
-            **kwargs
-        )
+        filtering_strings = [u"{}={}".format(key, value) for key, value in six.iteritems(filters)]
+
+        cache_key = u"tagstore.__get_tag_keys:{}".format(md5_text(*filtering_strings).hexdigest())
```

Comment:
I think you may end up into cache key size problems if you just concatenate all the filters. There are organizations that contain hundreds of projects and that query all of them. 
Also this approach means that we will cache an entry for all combinations of projects requested by the user while we could run the query grouping by project, cache one entry per project and compose the result from multiple cache keys when needed.

Reply:
I'm using `md5_text` here from a suggestion on another PR with caching to keep the cache key below the size limit.

Also to make sure I understand your suggestion, if the user's query is for `n` projects. Run the snuba query for each project, cache each result individually, and return the union of all the results?

Reply:
ok,
no, what I meant is to still run one query only but also group by project_id so as a result you have tags and projects but, thinking over it, this can backfire if the limit of projects in input is very high and the result size limit is very high, since the snuba query result will be huge.
probably this approach is better as a first approximation.

Reply:
@wmak  If `projects = [1,2,3]` in one request, and `projects = [3,1,2]` (different permutation) in another request, this would result in two distinct cache keys.

We may want to treat these as being identical request payloads.

Similarly for `environments` and `tags`.

Reply:
Ooh true, I'll sort the values there to keep the consistent regardless of the order they're passed in.